File: .skills/openclaw-skills/skills/neo-ava/sparker/src/core/openclaw-config.js
// Auto-detect and reuse OpenClaw's model provider configuration.
// Falls back to STP_* / SPARK_* env vars if openclaw config not found.
//
// This allows sparker to work out-of-the-box when distributed to other
// openclaw users — no extra env vars needed.

var fs = require('fs');
var path = require('path');

var _cache = null;

function loadOpenClawConfig() {
  if (_cache) return _cache;

  var home = process.env.HOME || process.env.USERPROFILE || '/root';
  var modelsPath = path.join(home, '.openclaw', 'agents', 'main', 'agent', 'models.json');
  var mainPath = path.join(home, '.openclaw', 'openclaw.json');

  var providers = null;
  var primaryModel = null;

  // Read main config for primary model setting
  try {
    var mainRaw = fs.readFileSync(mainPath, 'utf8');
    var mainCfg = JSON.parse(mainRaw);
    if (mainCfg.agents && mainCfg.agents.defaults && mainCfg.agents.defaults.model) {
      primaryModel = mainCfg.agents.defaults.model.primary || null;
    }
    if (mainCfg.models && mainCfg.models.providers) {
      providers = mainCfg.models.providers;
    }
  } catch (e) { /* skip */ }

  // Models.json has per-provider apiKeys — merge into providers
  try {
    var modRaw = fs.readFileSync(modelsPath, 'utf8');
    var modCfg = JSON.parse(modRaw);
    if (modCfg.providers) {
      if (!providers) providers = {};
      for (var name in modCfg.providers) {
        if (!providers[name]) providers[name] = {};
        Object.assign(providers[name], modCfg.providers[name]);
      }
    }
  } catch (e) { /* skip */ }

  if (!providers) {
    _cache = { providers: {}, primaryModel: null };
    return _cache;
  }

  _cache = { providers: providers, primaryModel: primaryModel };
  return _cache;
}

// Resolve LLM config: env vars > openclaw config
function resolveLLMConfig() {
  var endpoint = process.env.STP_FORGE_LLM_ENDPOINT
    || process.env.SPARK_PROMOTE_LLM_ENDPOINT
    || process.env.LLM_ENDPOINT;
  var apiKey = process.env.STP_FORGE_LLM_API_KEY
    || process.env.SPARK_PROMOTE_LLM_API_KEY
    || process.env.LLM_API_KEY;
  var model = process.env.STP_FORGE_LLM_MODEL
    || process.env.SPARK_PROMOTE_LLM_MODEL
    || process.env.LLM_MODEL;

  if (endpoint) {
    return { baseUrl: endpoint, apiKey: apiKey || '', model: model || 'gpt-4o-mini', api: 'openai' };
  }

  var cfg = loadOpenClawConfig();

  // Prefer the primary model's provider
  var primaryProvider = null;
  if (cfg.primaryModel) {
    var parts = cfg.primaryModel.split('/');
    if (parts.length >= 2) primaryProvider = parts[0];
  }

  var names = Object.keys(cfg.providers);
  if (primaryProvider && names.indexOf(primaryProvider) >= 0) {
    names = [primaryProvider].concat(names.filter(function (n) { return n !== primaryProvider; }));
  }

  for (var i = 0; i < names.length; i++) {
    var p = cfg.providers[names[i]];
    if (p.baseUrl && p.apiKey) {
      var modelId = (p.models && p.models.length > 0) ? p.models[0].id : 'default';
      return {
        baseUrl: p.baseUrl,
        apiKey: p.apiKey,
        model: modelId,
        api: p.api || 'openai',
        provider: names[i],
      };
    }
  }
  return null;
}

// Resolve Embedding config: env vars > openclaw config
function resolveEmbeddingConfig() {
  var endpoint = process.env.STP_EMBEDDING_ENDPOINT
    || process.env.SPARK_EMBEDDING_ENDPOINT;
  var apiKey = process.env.STP_EMBEDDING_API_KEY
    || process.env.SPARK_EMBEDDING_API_KEY;
  var model = process.env.STP_EMBEDDING_MODEL;

  if (endpoint) {
    return { endpoint: endpoint, apiKey: apiKey || '', model: model || 'default' };
  }

  var cfg = loadOpenClawConfig();
  var names = Object.keys(cfg.providers);

  // Prefer the primary provider for embeddings too
  var primaryProvider = null;
  if (cfg.primaryModel) {
    var parts = cfg.primaryModel.split('/');
    if (parts.length >= 2) primaryProvider = parts[0];
  }
  if (primaryProvider && names.indexOf(primaryProvider) >= 0) {
    names = [primaryProvider].concat(names.filter(function (n) { return n !== primaryProvider; }));
  }

  // Prefer doubao/volcengine for embeddings (has native embedding models)
  var ordered = names.slice();
  var doubaoIdx = ordered.indexOf('doubao');
  if (doubaoIdx > 0) {
    ordered.splice(doubaoIdx, 1);
    ordered.unshift('doubao');
  }

  for (var i = 0; i < ordered.length; i++) {
    var p = cfg.providers[ordered[i]];
    if (p.baseUrl && p.apiKey) {
      var base = p.baseUrl.replace(/\/$/, '');

      // Doubao / Volcengine ARK: use multimodal embedding endpoint
      if (base.indexOf('volces.com') >= 0 || ordered[i] === 'doubao') {
        base = base.replace(/\/api\/v3$/, '').replace(/\/v1$/, '');
        return {
          endpoint: base + '/api/v3/embeddings/multimodal',
          apiKey: p.apiKey,
          model: process.env.STP_EMBEDDING_MODEL || 'doubao-embedding-vision-251215',
          provider: ordered[i],
          derived: true,
        };
      }

      base = base.replace(/\/anthropic$/, '').replace(/\/v1$/, '');
      return {
        endpoint: base + '/v1/embeddings',
        apiKey: p.apiKey,
        model: 'default',
        provider: ordered[i],
        derived: true,
      };
    }
  }
  return null;
}

// Generic LLM call that handles both OpenAI and Anthropic formats
async function callLLM(prompt, opts) {
  var cfg = opts || resolveLLMConfig();
  if (!cfg) return null;

  var isAnthropic = cfg.api && cfg.api.indexOf('anthropic') >= 0;
  var url, headers, body;

  if (isAnthropic) {
    url = cfg.baseUrl.replace(/\/$/, '') + '/v1/messages';
    headers = {
      'Content-Type': 'application/json',
      'x-api-key': cfg.apiKey,
      'anthropic-version': '2023-06-01',
    };
    body = JSON.stringify({
      model: cfg.model,
      max_tokens: opts && opts.max_tokens || 2000,
      messages: [{ role: 'user', content: prompt }],
      temperature: opts && opts.temperature !== undefined ? opts.temperature : 0.3,
    });
  } else {
    url = cfg.baseUrl.replace(/\/$/, '');
    if (url.indexOf('/chat/completions') === -1) url += '/v1/chat/completions';
    headers = {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer ' + cfg.apiKey,
    };
    body = JSON.stringify({
      model: cfg.model,
      messages: [{ role: 'user', content: prompt }],
      temperature: opts && opts.temperature !== undefined ? opts.temperature : 0.3,
      max_tokens: opts && opts.max_tokens || 2000,
    });
  }

  try {
    var res = await fetch(url, {
      method: 'POST',
      headers: headers,
      body: body,
      signal: AbortSignal.timeout(60000),
    });
    var data = await res.json();

    if (isAnthropic) {
      var blocks = data.content || [];
      for (var i = 0; i < blocks.length; i++) {
        if (blocks[i].type === 'text') return blocks[i].text;
      }
      return null;
    } else {
      return data.choices && data.choices[0] && data.choices[0].message
        && data.choices[0].message.content;
    }
  } catch (e) {
    return null;
  }
}

module.exports = {
  loadOpenClawConfig,
  resolveLLMConfig,
  resolveEmbeddingConfig,
  callLLM,
};

File: .skills/openclaw-skills/skills/cwyhkyochen-a11y/devtaskflow/lib/openclaw_config.py
"""从 OpenClaw 配置自动读取 LLM 设置"""
import json
import os
from pathlib import Path


def _load_openclaw_config() -> dict:
    config_path = Path.home() / '.openclaw' / 'openclaw.json'
    if not config_path.exists():
        return {}
    try:
        return json.loads(config_path.read_text(encoding='utf-8'))
    except Exception as e:
        print(f'⚠️ openclaw_config: 读取配置文件失败: {e}')
        return {}


def _read_api_key(provider: str) -> str:
    """从 ~/.openclaw/credentials/ 读取 API Key"""
    cred_dir = Path.home() / '.openclaw' / 'credentials'
    # 尝试 {provider}:default.api
    for suffix in [f'{provider}:default.api', f'{provider}.api']:
        cred_file = cred_dir / suffix
        if cred_file.exists():
            try:
                return cred_file.read_text(encoding='utf-8').strip()
            except Exception as e:
                print(f'⚠️ 读取凭据文件 {cred_file} 失败: {e}')
                continue
    return ''


def detect_openclaw_llm() -> dict:
    """自动检测 OpenClaw 的 LLM 配置，返回 {base_url, api_key, model}。
    任一字段获取失败则对应值为空字符串。
    """
    cfg = _load_openclaw_config()
    if not cfg:
        return {'base_url': '', 'api_key': '', 'model': ''}

    # 1. 获取主模型 (如 "openrouter/xiaomi/mimo-v2-pro")
    primary = cfg.get('agents', {}).get('defaults', {}).get('model', {}).get('primary', '')
    if not primary:
        return {'base_url': '', 'api_key': '', 'model': ''}

    # 2. 解析 provider 名
    # 格式: "provider/model-id" 或 "provider/sub-provider/model-id"
    parts = primary.split('/')
    providers = cfg.get('models', {}).get('providers', {})

    # 3. 在 providers 中查找匹配的 baseUrl
    base_url = ''
    api_key = ''
    model_id = parts[-1] if parts else primary

    # 尝试从最长到最短的 provider 名匹配
    for i in range(len(parts) - 1, 0, -1):
        provider_name = '/'.join(parts[:i])
        if provider_name in providers:
            base_url = providers[provider_name].get('baseUrl', '').rstrip('/')
            api_key = _read_api_key(provider_name)
            break

    # 如果没找到，尝试第一个部分作为 provider
    if not base_url and parts and parts[0] in providers:
        base_url = providers[parts[0]].get('baseUrl', '').rstrip('/')
        api_key = _read_api_key(parts[0])

    return {
        'base_url': base_url,
        'api_key': api_key,
        'model': model_id or primary,
    }
